### DISCUSSION TITLE
Semi Pie Chart taking up to much space?

### DISCUSSION THREAD

**Discussion by yusuf8ahmed** [recharts/recharts#3846] [Q&A] [answered]
Hi Guys, 
I am currently making a dashboard, because i want to learn css flexbox, grid, tailwind and jsx. The current problem is that i want to put buttons under the  PieChart but there is too much space for some reason?

![Screenshot 2023-10-11 at 3 36 30 PM](https://github.com/recharts/recharts/assets/33407772/7b58f8c4-a9e3-44be-9d83-6e938787cf6f)

My grid tailwindcss classNames are:
grow h-full grid grid-cols-3 grid-rows-4 auto-rows-min sm:grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-4 items-stretch

My tailwindcss classNames for this box is:
flex flex-col row-span-1 rounded-xl border-1 border-slate-400/10 bg-gray-100 p-4 dark:bg-slate-900 lg:row-span-2

And this is my PieChart inside the box above:
```
<div className="flex flex-row items-center justify-between flex-none">
  <div className="flex flex-row items-center">
    <CreditScoreIcon />  
    <p className="m-0 ml-1">Spending Summary</p>
  </div>

  <Select defaultValue="week">
    <SelectTrigger className="ml-auto max-w-max pr-2">
      <SelectValue placeholder="This Week"/>
    </SelectTrigger>
    <SelectContent>
      <SelectItem value="day">Today</SelectItem>
      <SelectItem value="week">This Week</SelectItem>
      <SelectItem value="month">Last 30 days</SelectItem>
    </SelectContent>
  </Select>
</div>
<ResponsiveContainer width="100%" className="min-h-0 shrink" aspect={1.2}>
  <PieChart margin={{ top: 0, left: 0, right: 0, bottom: 0 }} fill="teal">
    <Pie
      data={pieChartData}
      startAngle={180}
      endAngle={0}
      innerRadius="60%"
      outerRadius="80%"
      fill="#8884d8"
      dataKey="value"
      // blendStroke
      // stroke={{opacity: 1}}
    >
      {pieChartData.map((_, index) => (
        <Cell key={`cell-${index}`} fill={COLORS[index % COLORS.length]} />
      ))}
    </Pie>
  </PieChart>
</ResponsiveContainer>
```

example below when i put a separator (br element) after the separator 
 ![Screenshot 2023-10-11 at 3 41 06 PM](https://github.com/recharts/recharts/assets/33407772/fabd178d-d164-43f8-8550-485033da9a78)

**Comment by ckifer** [comment]
hey @yusuf8ahmed, this is going to be difficult to help with without a reproduction. Could you make a minimal codesandbox?

**Reply by yusuf8ahmed** [reply]
Hi ckifer,

https://codesandbox.io/p/sandbox/eloquent-hopper-lmrfmz?file=%2Fapp%2F(dashboard)%2Fhome%2F_components%2Fdashboard%2Findex.js%3A1%2C1

The route is[ /home ](https://lmrfmz-3003.csb.app/home)
and the file is [app/(dashboard)/home/_components/dashboard/index.js](https://codesandbox.io/p/sandbox/eloquent-hopper-lmrfmz?file=%2Fapp%2F(dashboard)%2Fhome%2F_components%2Fdashboard%2Findex.js%3A1%2C1)

**Reply by ckifer** [reply]
Hmm, the sandbox isn't working for me

**Reply by yusuf8ahmed** [reply]
what is it showing on your side?

**Reply by ckifer** [reply]
The code is there it just won't run, I get some random build failures
```
[CODESANDBOX] .codesandbox/Dockerfile detected
[CODESANDBOX] Waiting for container build to finish...
Failed to run containers exec run: Waiting for condition timed out
```

I might have to download it

**Reply by ckifer** [reply, marked as answer]
Okay I pulled this down and took a look.

Basically recharts piechart does not currently adjust when the size of the chart itself changes. The size of the screen yes, but since you are only rendering 180 to 0 (half a circle) the container is still assuming you are taking up double that space.

commenting out your startAngle/endAngle makes that relatively evident. 

If you want you can create a separate issue that details this, sorry for the inconvenience!

A hacky way to get around it for now is to add a whole lot of negative bottom margin (like -250)
![image](https://github.com/recharts/recharts/assets/25180830/d3d809db-7db1-4e3d-ae42-3f857d4f60e5)

**Reply by yusuf8ahmed** [reply]
Is there a way that i could make it responsive (instead of -250px)?, and i wanted to fix recharts, to start recognizing that the PieChart is half height, where should i start in the repo?

also i really appreciate all the help, it is open source maintainers like you that keep the spirit of collaboration and community alive.

**Reply by ckifer** [reply]
With should be responsive with your container. Height however will always need to be fixed regardless if this is working or not. 

Sure, I can take a look some point.  Will get back in the next few days

@nikolasrieble  if you have any ideas of where this could be done

**Reply by nikolasrieble** [reply]
While this is easy for a single Pie in a PieChart, how about multiple Pies? Each have their own start and endAngle. We would have to aggregate all start and endAngles across the Pies, and then render them accordingly. It seems that they are currently rendered independently. 

We could expose the startAngle and endAngle on the PieChart API (as opposed to the Pie API). 

## Another solution

Alternative to a negative margin, you could use `cy` to set the relative center y of the Pie. 

```
export default function App() {
  return (
    <PieChart width={400} height={400}>
      <Pie
        data={data02}
        dataKey="value"
        innerRadius={70}
        outerRadius={90}
        startAngle={120}
        endAngle={180}
        cy="80%"
      />
    </PieChart>
  );
}
```

**Reply by ckifer** [reply]
Thanks Niko. Yeah cy might be a better solution here, forgot about that. 

Imo a PieChart should only ever have one Pie in it, if it has more id recommend creating a separate PieChart. 
Ideally the content in the fixed height container fills that height with little or no empty space when margin is 0 (easier said than done of course)

**Reply by nikolasrieble** [reply]
Multi-level PieCharts require multiple Pies though, and that is a great chart / use case.

**Reply by yusuf8ahmed** [reply]
@nikolasrieble if i wanted to contribute to fixing this issue what file should I start with (where is the file the sets the height for Pies?), i saw Surface and Layer but i don't really understand how it all connects with Pie and PieChart component?

**Reply by nikolasrieble** [reply]
Surface and Layer are generic parent components, used across all charts. Chances are high the the generateCategoricalChart file will be touched 😅

**Reply by ckifer** [reply]
Honestly the `cy` solution here is pretty ideal. `75%` gets you the middle every time. If we change this default behavior we also have to account for it no matter the value of `cy` - maybe we should only do anything here if the value of `cy` is not provided.

@yusuf8ahmed you'll essentially have to modify the `cy` value yourself for each sector 
https://github.com/recharts/recharts/blob/master/src/polar/Pie.tsx#L499.

But not only for each sector, for everything in the Pie (labels, etc.)

**Reply by murray-b** [reply]
Following up here because I was facing a similar issue -- how does `cy` solve the issue described here? All that does is shift the chart down, but doesnt solve the issue that its container is much too large.

**Reply by ckifer** [reply]
@murray-b feel free to open an issue if the cy approach doesn't help. Agreed it still takes up too much space

**Reply by Morten-Moeller** [reply]
I used cy="100%" and outerRadius="200%" to position the pie chart’s center at the bottom. Since we’re only displaying a half pie, this allows us to double its size effectively.

**Reply by nmashchenko** [reply]
200% makes radius really weird I d say pretty unusable @Morten-Moeller the only real solution that helped was adding margin but this doesn't really work well
